src/util/locking/WrappedExpiringResourceLocker.ts

```diff
@@ -24,14 +23,62 @@ export class WrappedExpiringResourceLocker implements ExpiringResourceLocker {
     this.expiration = expiration;
   }
 
+  public async withReadLock<T>(identifier: ResourceIdentifier, func: (renew: () => void) => T | Promise<T>):
+  Promise<T> {
+    return this.locker.withReadLock(identifier, async(): Promise<T> => this.timePromise(identifier, func));
+  }
+
+  public async withWriteLock<T>(identifier: ResourceIdentifier, func: (renew: () => void) => T | Promise<T>):
+  Promise<T> {
+    return this.locker.withWriteLock(identifier, async(): Promise<T> => this.timePromise(identifier, func));
+  }
+
+  /**
+   * Creates a Promise that either resolves the given input function or rejects if time runs out,
+   * whichever happens first. The input function can reset the timer by calling the `renew` function
+   * it receives. The ResourceIdentifier is only used for logging.
+   */
+  private async timePromise<T>(identifier: ResourceIdentifier, func: (renew: () => void) => T | Promise<T>):
+  Promise<T> {
+    let timer: Timeout;
+    let timerReject: (error: Error) => void;
+
+    // Promise that throws an error when the timer finishes
+    const timerPromise = new Promise<never>((resolve, reject): void => {
+      timerReject = reject;
+      timer = this.resetTimer(identifier, timerReject);
+    });
+
+    // Restarts the timer
+    const refresh = (): void => {
+      this.logger.verbose(`Renewed expiring lock on ${identifier.path}`);
+      timer = this.resetTimer(identifier, timerReject, timer);
+    };
+
+    // Runs the main function and cleans up the timer afterwards
+    async function resolveFunc(): Promise<T> {
+      try {
+        return await func(refresh);
+      } finally {
+        clearTimeout(timer);
+      }
+    }
+
+    return Promise.race([ timerPromise, resolveFunc() ]);
+  }
+
   /**
-   * Lock the given resource with a lock providing expiration functionality.
-   * @param identifier - Identifier of the resource that needs to be locked.
-   *
-   * @returns A promise containing the expiring lock on the resource.
+   * Creates a new TimeOut that will call the given reject function once time runs out.
+   * Clears the given input Timeout if one is provided.
+   * The ResourceIdentifier is only used for logging.
    */
-  public async acquire(identifier: ResourceIdentifier): Promise<ExpiringLock> {
-    const innerLock = await this.locker.acquire(identifier);
-    return new WrappedExpiringLock(innerLock, this.expiration);
+  private resetTimer(identifier: ResourceIdentifier, reject: (reason: Error) => void, timer?: Timeout): Timeout {
```

Comment:
It's weird to see this as a member function; the only `this` it uses is the logger (so not really). Shall we have it inside of the `timePromise` function as a closure?

Reply:
Only did this to keep the `timePromise` function small and readable, I'll move it.